src/Compilers/CSharp/Test/Emit3/RefStructInterfacesTests.cs

```diff
@@ -11182,92 +11446,232 @@ static void Main()
-    }
+    public int Current => throw null;
+    public bool MoveNext() => throw null;
+    public void Dispose() => throw null;
 }
+";
+
+            var comp1 = CreateCompilation(src1, targetFramework: s_targetFrameworkSupportingByRefLikeGenerics);
 
+            var src2 = @"
 class C
 {
     static void Main()
     {
         foreach (var i in new S1())
         {
-            System.Console.Write(i);
         }
     }
 }
 ";
-            var comp = CreateCompilation(src, targetFramework: s_targetFrameworkSupportingByRefLikeGenerics, options: TestOptions.ReleaseExe);
+            var comp2 = CreateCompilation(src2, references: [comp1.ToMetadataReference()], targetFramework: s_targetFrameworkSupportingByRefLikeGenerics, parseOptions: TestOptions.Regular12);
+            comp2.VerifyEmitDiagnostics();
 
-            CompileAndVerify(
-                comp, expectedOutput: ExecutionConditionUtil.IsMonoOrCoreClr ? @"123D" : null,
-                verify: ExecutionConditionUtil.IsMonoOrCoreClr ?
-                    Verification.FailsILVerify with { ILVerifyMessage = "[GetEnumerator]: Return type is ByRef, TypedReference, ArgHandle, or ArgIterator. { Offset = 0x9 }" } :
-                    Verification.Skipped).VerifyDiagnostics();
+            comp2 = CreateCompilation(src2, references: [comp1.ToMetadataReference()], targetFramework: s_targetFrameworkSupportingByRefLikeGenerics, parseOptions: TestOptions.RegularNext);
+            comp2.VerifyEmitDiagnostics();
+
+            comp2 = CreateCompilation(src2, references: [comp1.ToMetadataReference()], targetFramework: s_targetFrameworkSupportingByRefLikeGenerics);
+            comp2.VerifyEmitDiagnostics();
         }
 
         [Fact]
-        public void Foreach_Pattern_02()
+        public void Foreach_IDisposable_LanguageVersion_03()
         {
-            var src = @"
-struct S1
+            var src1 = @"
+public struct S1
 {
     public S2 GetEnumerator()
     {
         return new S2();
     }
 }
 
-ref struct S2
+public ref struct S2 : System.IDisposable
 {
-    bool stop;
-    public int Current => 123;
-    public bool MoveNext()
-    {
-        if (!stop)
-        {
-            stop = true;
-            return true;
-        }
-
-        return false;
-    }
+    public int Current => throw null;
+    public bool MoveNext() => throw null;
+    void System.IDisposable.Dispose() => throw null;
 }
+";
+
+            var comp1 = CreateCompilation(src1, targetFramework: s_targetFrameworkSupportingByRefLikeGenerics);
 
+            var src2 = @"
 class C
 {
     static void Main()
     {
-        foreach (var i in new S1())
-        {
-            System.Console.Write(i);
+        foreach (var i in new S1()) {}
+    }
+}
+";
+            var comp2 = CreateCompilation(src2, references: [comp1.ToMetadataReference()], targetFramework: s_targetFrameworkSupportingByRefLikeGenerics, parseOptions: TestOptions.Regular12);
+            comp2.VerifyEmitDiagnostics(
```

Comment:
>VerifyEmitDiagnostics [](http://example.com/codeflow?start=18&length=21)

Why do we report an error here for `-langversion:12` using `IDisposable`, but we don't report an error for the `-langversion:12` case using `IEnumerable<int>` in `Foreach_IEnumerableT_LanguageVersion_03`?

Reply:
> Why do we report an error here for `-langversion:12` using `IDisposable`, but we don't report an error for the `-langversion:12` case using `IEnumerable<int>` in `Foreach_IEnumerableT_LanguageVersion_03`?

Semantic analysis didn't rely and still doesn't rely on conversion classification for generic IEnumerable case. I will check behavior in main if it runs into problems elsewhere and will follow up separately, if needed.

Reply:
Besides, `Foreach_IEnumerableT_LanguageVersion_03` doesn't dispose a ref struct